rest-client/src/main/java/org/jboss/pnc/restclient/websocket/VertxWebSocketClient.java

```diff
@@ -0,0 +1,317 @@
+/**
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2014-2019 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.pnc.restclient.websocket;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+
+import org.jboss.pnc.common.json.JsonOutputConverterMapper;
+import org.jboss.pnc.dto.notification.BuildChangedNotification;
+import org.jboss.pnc.dto.notification.BuildConfigurationCreation;
+import org.jboss.pnc.dto.notification.BuildPushResultNotification;
+import org.jboss.pnc.dto.notification.GroupBuildChangedNotification;
+import org.jboss.pnc.dto.notification.Notification;
+import org.jboss.pnc.dto.notification.RepositoryCreationFailure;
+import org.jboss.pnc.dto.notification.SCMRepositoryCreationSuccess;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import io.vertx.core.Vertx;
+import io.vertx.core.http.HttpClient;
+import io.vertx.core.http.WebSocket;
+
+/**
+ * @author <a href="mailto:jmichalo@redhat.com">Jan Michalov</a>
+ */
+public class VertxWebSocketClient implements WebSocketClient, AutoCloseable {
+
+    private static final Logger log = LoggerFactory.getLogger(VertxWebSocketClient.class);
+
+    private static final ObjectMapper objectMapper = JsonOutputConverterMapper.getMapper();
+
+    private final Vertx vertx;
+
+    private HttpClient httpClient;
+
+    private WebSocket webSocketConnection;
+
+    private Set<Dispatcher> dispatchers = new HashSet<>();
+
+    private Set<CompletableFuture<Notification>> singleNotificationFutures = new HashSet<>();
+
+    /**
+     * number of attempted retries before client throws an exception
+     */
+    private int maximumRetries = 5;
+
+    private int numberOfRetries = 0;
+
+    /**
+     * a multiplier that increases delay between reconnect attempts
+     */
+    private float delayMultiplier = 1.5F;
```

Comment:
I'd say just use as delayMultiplier the number '2'.

And with that you can optimize by doing a bitwise shift left to "multiply" by 2.

Reply:
For sure can change it to 2. With bitwise shift it forces the number to be 2 and doesn't allow for the number to change (currently can be set through constructor). Do we want to hard code it?

Reply:
I once heard some thing about premature optimization. Doesn't it apply here?

Reply:
errm, changing from 1.5 to 2 is a pretty big increase in backoff factor... isn't 1.3 usually the standard?

Reply:
actually I don't know where I got 1.3 from, my brain offered it up, but I can't find anything to back it up...

Reply:
I've made that number up, so I am free to change it to whatever you want. We can also make it customizable through CLI config or smthing like that.